### ISSUE TITLE
Better hints, custom layouts

### ISSUE THREAD

**Issue by Morantron** [Morantron/tmux-fingers#16]
# Problem

As pointed out [here](https://github.com/Morantron/tmux-fingers/pull/15#issuecomment-255548758), right now hints are hardcoded in [hinter.awk](https://github.com/Morantron/tmux-fingers/blob/7c8356dc07dac3418073cd38d2c862e4fa8eead0/scripts/hinter.awk#L5), and they can be improved.

The main problem is that one hint cannot be a subtset of another one: if you have a hint `a` you cannot have `as` `ad` or `as`, other hints simply cannot start by `a`.

Hints to be used depends on the number of matches. Imagine your alphabet is `asdf`.
- For 3 matches you should use: `a`, `s`, `d`.
- For 5 matches you should use: `s` `d` `f` `aa` `as`

As you can see, none of the the hints is a subset of another one.

This problem is solved by the _n-ary huffman algorithm_  ( more links below ).

The harcoded hints were calculated using a small throwaway node script that uses this algorithm. However, I set up a number of 100 hints, making most of the hints to be two characters long, even if they are not necessary, and making most of the one-char hints out of the home row ( of a qwerty layout :trollface: ).
# Solution

Implementing n-ary-huffman in bash/awk/gawk :skull: 

What would be improved:
- How hints are assigned, making them longer than one char only if necessary.
- User configurable sequence ( qwerty, vim-sneak, dvorak ).
- Most of the hints would fall under the user's home row, or very close.

However, we should take into account:
- Performance should not be affected. Last time I measured, the time it takes for hints to render is near **15-20ms**. If the n-ary-huffman implementation is slow, its results could be precomputed cached the first time fingers is invoked for a prefixed amount of hints ( 5, 10, 25, or everthing between 1-100 ).
- There should be no external dependencies ( or _no more_ if you consider gawk one ). Making tree-like structures in a language like this will be certainly challenging and fun. :trollface: If hints are precomputed maybe something like ( https://github.com/kanaka/mal ) could be used and the implementation could be easier.

Huffman related links:

https://en.wikipedia.org/wiki/Huffman_coding
https://www.npmjs.com/package/n-ary-huffman - npm package used to calculate hardcoded strings
https://rosettacode.org/wiki/Huffman_coding - reference implementations in lots of languages

**Comment by sunaku** [comment]
If I may humbly suggest, I would approach this problem in two steps: :feet:

* :hocho: Eliminate the complexity of how to choose hints by moving that decision _outside_ of this plugin: make hints configurable via a new `@fingers-hint-keys` tmux configuration setting, which internally transforms into an environment variable when dealing with the AWK hinter. :scissors:

* :books: Provide examples of configuring the `@fingers-hint-keys` setting for qwerty, vim-sneak, dvorak, etc.  :rocket: Here, you can go all out, choosing sophisticated algorithms to pre-compute hints on behalf of users, who will then choose/copy/paste/modify these snippets out of the README.

* Keep the current hard-coded set of hints as a fallback, in case the user doesn't configure anything. :sleeping:

What do you think? :neckbeard:

**Comment by Morantron** [comment]
Interresting.

Not sure about moving the decision outside of the plugin, there is a difference between _alphabet_ and _hint_. An alphabet produces a set hints for a number of matches, and that produced hints cannot intersect between them.

```
alphabet(3) = ['a', 's', 'd']
alphabet(5) = [''s', 'd', 'f', 'aa', 'as']
```

Maybe that's too advanced so I think it's better if the users should be able to define an _alphabet_ instead of the hints per se.

An easier approach like you pointed out could be: precomputing hints for a predefined set of alphabets with an external tool. So users would be able to decide for example between `qwerty`, `dvorak` or `sneak`.

`set @fingers-layout -g "qwerty"`

Then later, ideally, with n-ary-huffman implemented with no external dependencies ( for science ), we could precompute hints for any alphabet and store them in a folder somewhere.

Thoughts?

**Comment by sunaku** [comment]
Alright, since you're so keen on implementing that n-arry-huffman algorithm. :smirk_cat:

But I would prefer to define the actual hints (not alphabet, and not layout) myself:

```sh
set @fingers-hint-labels " a o e u i d h t n s ' , . p y f g c r l / ; q j k x b m w v z A O E U I D H T N S \ " < > P Y F G C R L ? : Q J K X B M W V Z 1 2 3 4 5 6 7 8 9 0 "
```

This example specifies 72 distinct hints.  If the hinter finds more than 72 matches, the extra ones will just remain unselectable (which is the current behavior) and, for the time being, that's acceptable to me. :face_with_head_bandage:

The additional `@fingers-hint-alphabet` setting can build on top of this `@fingers-hint-labels` setting.  Similarly, the `@fingers-hint-layout` setting can build on top of `@fingers-hint-alphabet`.

Basically, I'm trying to attack the underlying problem of hard-coded hints with minimum code changes. :neckbeard:

**Comment by sunaku** [comment]
Well, I've put my code :money_mouth_face: where my mouth is, so to speak, by submitting PR #17. :gift:  Coding beats talking.

**Comment by dragon788** [comment]
@sunaku This sounds awesome. One of the things that I really like being able to do in Vimium for example is define that link hints can be drawn only from the left hand side of the keyboard (since I'm mousing with my right hand), so if alphabet or layout is qwerty, `asdfg qwerty zxcvb` though I have a preference for the first two groups, aka home row and top row (but as many characters as necessary). Is this something your hint labels will accomplish, and then as you said I could define an alphabet that would called "left hand" that encompasses those characters?

Really I'd love to see what a combination of the solutions turns into. As a combination of the above it seems like labels could be "let me override the characters I want offered as shortcuts", with some pre-defined macros like "left hand", "homerow", "numbers","easy rows" (aka home row and above).

I then kind of imagine alphabet as "all possible characters", then keymap may narrow it to type-able characters, so on a US/ANSI keyboard this doesn't have umlaut or the various European accent characters but if defined as ISO it offers those as hints, then layout is a hint to the program "where home row lives" eg qwerty/azerty/dvorak/colemak. This would allow for unset values inheriting from the OS mappings, but a user could override them if necessary.

**Comment by Morantron** [comment]
Available in 1.0.0. Check https://github.com/morantron/tmux-fingers#fingers-keyboard-layout
